lib/puppet/pops/types/type_calculator.rb

```diff
@@ -1192,27 +1200,44 @@ def tuple_entry_at(tuple_t, from, to, index)
   # @api private
   #
   def assignable_PStructType(t, t2)
-    return true if t == t2 || t.elements.empty? && (t2.is_a?(Types::PHashType))
-    h = t.hashed_elements
+    elements = t.elements
     if t2.is_a?(Types::PStructType)
       h2 = t2.hashed_elements
-      h.size == h2.size && h.all? {|k, v| assignable?(v, h2[k]) }
+      match_count = 0
+      ok = elements.reduce(true) do |memo, e|
+        break false unless memo
+        t2e = h2[e.name]
+        if t2e.nil?
+          is_type_optional?(e.type)
+        else
+          match_count += 1
+          assignable?(e.type, t2e)
+        end
+      end
+      ok && match_count == h2.size
```

Comment:
That was the real problem then? instead of using h2[k] directly, it should have used a PNilType. The rest is the same change as above that does not use all?